### DISCUSSION TITLE
RFC - Automated recovery when E2E fail in production (CI/CD)

### DISCUSSION THREAD

**Discussion by Vadorequest** [UnlyEd/next-right-now#48] [Ideas]
Goal: Discuss desired behaviour about automated rollback to the "previously successful" deployment when a new deployment fails E2E tests in production stage.

> Quick question, I've noticed that the E2E CI process is running after the Deployment, and sometimes the deployment might succeed even though the e2e might fail, which is weird, I think the correct approach would be run any tests before the deployment, and do not deploy if it fails. - @samuelcastro

**Comment by samuelcastro** [comment]
We could use a [blue-green strategy](https://docs.cloudfoundry.org/devguide/deploy-apps/blue-green.html), I personally used it in project a while ago (not a next.js project), it's a robust solution, the problem I see with just rolling back when e2e tests fails is that production will have a bad deployment live for quite some time (the time to finish up e2e).

> Blue-green deployment is a technique that reduces downtime and risk by running two identical production environments called Blue and Green.
> 
> At any time, only one of the environments is live, with the live environment serving all production traffic. For this example, Blue is currently live and Green is idle.
> 
> As you prepare a new version of your software, deployment and the final stage of testing takes place in the environment that is not live: in this example, Green. Once you have deployed and fully tested the software in Green, you switch the router so all incoming requests now go to Green instead of Blue. Green is now live, and Blue is idle.

**Comment by samuelcastro** [comment]
More relevant links:

https://martinfowler.com/bliki/BlueGreenDeployment.html
https://spectrum.chat/zeit/now/blue-green-strategy~d1e657ae-761d-4668-9ed6-8f1648cf358e
https://github.com/zeit/next.js/issues/4337

**Comment by Vadorequest** [comment]
The design is very interesting, but the implementation is non-trivial, really.

Since Vercel only allows one production deployment per project, this would mean it'd switch between 2 different projects (blue/green), there is a lot of stuff behind such automation. And it'd be a very different implementation depending on the hosting provider.

**Comment by samuelcastro** [comment]
I think just by using alias would make things easier see: https://spectrum.chat/zeit/now/blue-green-strategy~d1e657ae-761d-4668-9ed6-8f1648cf358e

**Comment by Vadorequest** [comment]
Indeed, that may simplify things quite a lot.
